Change tracking

kravets-levko opened the pull request:
## What type of PR is this? (check all applicable)

- [x] Refactor
- [x] Feature

## Description

- [x] Improve existing change tracking mechanism:
  - [x] choose which attributes to track
  - [x] track creations and deletions (in addition to modifications)
  - [x] track relations (visualizations for queries, widgets for dashboard)
    - [x] when forking query, log creation of all new visualizations
    - [x] track API key updates
- [x] Enable change tracking for models:
  - [x] Query
    - [x] Visualization
  - [x] Dashboard
    - [x] Widget
    - [x] ApiKey
  - [x] Alert
  - [x] NotificationDestination
- [x] API to get recent changes
- [x] Migration to clear old data and update table structure
- [x] Update tests
  - [x] for models
  - [x] for API endpoints

## Example of data stored in DB

![image](https://user-images.githubusercontent.com/12139186/78550934-9badfb80-780d-11ea-8266-8c1aaf5ddffe.png)

Closes #3136.

kravets-levko commented:
@arikfr (cc @rauchy) I still need to update tests, but the feature code is ready and I'd like to know your opinion about the result/what could be improved/etc. Also, I have two more questions:

1. when forking query, should it also log creation of all new visualizatons?
2. should it track API key updates (for dashboards and queries)?

arikfr commented on redash/handlers/dashboards.py:
```diff
@@ -107,6 +107,10 @@ def post(self):
         models.db.session.commit()
+
+        dashboard.record_changes(self.current_user, models.Change.Type.Created)
+        models.db.session.commit()
```
Why commit twice? If it's because the object doesn't exist for the reference in changes -- SQLA knows to take care of this and supposed to save them in order and update the id. Did you try it?

kravets-levko replied:
Tried, and without this `commit()` I get this error: `sqlalchemy.exc.IntegrityError: (psycopg2.errors.NotNullViolation) null value in column "object_id" violates not-null constraint`. So I guess it doesn't work (maybe in some cases SQLA can handle this, but in this particular case it didn't)

kravets-levko replied:
It makes sense, but it will be harder to add change tracking to new models (e.g. Alerts). I think properly configured indexes should work fine as well

kravets-levko replied:
> Maybe because we're using non "traditional" foreign key it fails.

I've read some docs about SQLA and how it handles data changes, so I guess I know why it happens. When instance it created - it lives on its own until added to any session. When `sesson.commit()` is called - SQLA takes all new/dirty/removed instanced associated with the session, creates a SQL query for each of them, executes and only after that updates instances. In a case with new instances that means that `id` will be populated only _after_ `session.commit()`. I think it has some special handling for a case when there is a properly set-up relation between models, but `GFKBase` does not have relations in SQLA's meaning, therefore SQLA doesn't know that `Change` instances depend on `Query`/`Dashboard`/etc. being saved within the same `session.commit()`

arikfr commented on redash/models/changes.py:
```diff
@@ -1,12 +1,20 @@
 
 
 @generic_repr("id", "object_type", "object_id", "created_at")
 class Change(GFKBase, db.Model):
```
We need to add index(es) for this table to make it efficient to run a query like: `give me last N changes for object type X and id Y`.

kravets-levko replied:
Currently I added only index for `object_type, object_id` pair. When I'll start working on API - I'll check if we need more indexes or not.

arikfr commented:
> when forking query, should it also log creation of all new visualizatons?

Yes.

> should it track API key updates (for dashboards and queries)?

Yes.

kravets-levko commented on redash/models/changes.py:
```diff
@@ -46,48 +54,85 @@ def last_change(cls, obj):
+            return
+
+        changes = _get_object_changes(self)
+        # for `created` and `deleted` log even empty changes set
```
I think yes, but I'll check and ensure it if not

kravets-levko replied:
Done

kravets-levko commented on redash/models/changes.py:
```diff
@@ -46,48 +54,85 @@ def last_change(cls, obj):
+            user=changed_by,
+            change={
+                "object_type": self.__table__.name,
+                "object_id": self.id,
```
> Or you have it here for cases when changes.object is the parent?

Exactly. When we modify, say, visualization - we log it as a change to corresponding query; but then there is no other way to know that changes were made to visualization. I tried few different data structures, but this one looked the most straightforward

kravets-levko replied:
I added comment in the code that describes a structure of `changes` record and why we need `object_type/object_id` in `changes` field

kravets-levko commented on redash/models/changes.py:
```diff
@@ -1,12 +1,20 @@
+    class Type(str, Enum):  # `str` to make it json-serializable
+        Created = "created"
+        Modified = "modified"
+        Deleted = "deleted"
```
When object is archived - it's logged as a change to `is_archived` attribute. But if you think we need special handling for it - I can add it.

kravets-levko commented:
Recent changes (since last code review):

- when creating/forking query it now logs its visualizations as well;
- stopped logging `object_version`, removed this column from the `changes` table; updated migration script;
- added index for `object_type` + `object_id` fields;
- when creating new instance - log all initial fields (not just modified);
- log changes to API keys;
- optimized a way we access "parent" object - now it will not load instance from DB (this also simplified API keys tracking).

kravets-levko commented on redash/models/changes.py:
```diff
@@ -46,48 +54,85 @@ def last_change(cls, obj):
+                self.__changes__ = {}
+            change = self.__changes__.get(key, {
+                "previous": getattr(self, key),
+                "current": None,
```
Done. Also, I added a comment (in the code) that describes data structure

kravets-levko commented on redash/models/changes.py:
```diff
@@ -46,48 +54,85 @@ def last_change(cls, obj):
+
+        changes = Change(
+            object=self,
+            object_version=getattr(self, "version", None),
```
Done

kravets-levko commented on redash/models/changes.py:
```diff
@@ -46,48 +54,85 @@ def last_change(cls, obj):
-            previous = getattr(self, attr.key, None)
-            self._clean_values[col.name] = previous
+        if parent_attr:
+            changes.object = getattr(self, parent_attr, self)
```
Done. Also, I updated migration to remove this column from `changes` table on upgrade (and restore on downgrade).

kravets-levko commented on redash/models/changes.py:
```diff
@@ -46,48 +54,85 @@ def last_change(cls, obj):
+def track_changes(attributes, parent_attr=None):
+    attributes = set(attributes) - {"id", "created_at", "updated_at", "version"}
+
+    # monkey-patch class because inheritance will break SQLAlchemy
```
I decided to create a decorator for change tracking because it looks nice when you use in when defining a model. Decorator needs to modify original model class: intercept `__setattr__` and add a method to save changes. First I tried to inherit from original model class (`class ChangeTracking(cls)`) and it even worked. But in some cases SQLA was throwing errors - something like `class is not mapped` or like this (in particular that happened in tests). So eventually I checked how they implement own decorators in SQLA (e.g. `sqlalchemy_utils.models.generic_repr`) and realized that instead of using inheritance, they just monkey-patch the class (replace existing methods with own ones). I tried to do so - and it worked. OOP in Python is a bit weird; I guess when I inherited my class - it changed something in a mechanism SQLA uses to get model's metadata; when patching class I in fact keep the class itself untouched, and everything works

kravets-levko commented on migrations/versions/4952e040e9dd_clear_changes_table.py:
```diff
@@ -0,0 +1,34 @@
+depends_on = None
+
+
+def upgrade():
```
Done :+1:

jhult commented:
This looks like a cool feature. Any news on it?

guidopetri commented:
@kravets-levko , thanks for the PR! We've updated a lot of things now that we're Community-driven so - if you're still interested (or @jhult ) in getting this merged - would you mind rebasing off master to re-run the CI, as well as updating merge conflicts?

We're trying to clean up our PR todo list, so if you're not interested, that's fine - we'll close the PR in about a week if we don't hear back. If you're interested in reopening the PR afterwards, we would also very much welcome that.
